### PR TITLE
Add Rule keyword support

### PR DESCRIPTION
**Pull request by unkind**
Hi,

PR provides the `Rule` keyword support (https://cucumber.io/docs/gherkin/reference/#rule).

### CODE REVIEW

**Comment by ciaranmcnulty** [comment]
I really want this feature, looking forward to reviewing it

**Comment by unkind** [comment]
This one still doesn't work: `'complex_background.feature' => 'Rule keyword not supported',`.

It turns out `Scenario` and `Scenario Outline` have the same node type in Cucumber:

```gherkin
Feature: Complex background
...

  Rule: My Rule
    Scenario: with examples
      Given the <value> minimalism

      Examples:
      | value |
      | 1     |
      | 2     |
```

File: src/Behat/Gherkin/Keywords/KeywordsInterface.php [hunk 1]
```diff
@@ -100,4 +100,12 @@ public function getButKeywords();
...
      */
     public function getStepKeywords();
+
+
+    /**
+     * Returns all rule keywords (splitted by "|").
+     *
+     * @return string
+     */
+    public function getRuleKeywords();
```
Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Keywords/KeywordsInterface.php:110]: [BC] We should define a new interface that extends KeywordsInterface and adds this method

File: src/Behat/Gherkin/Node/FeatureNode.php [hunk 2]
```diff
@@ -72,6 +77,7 @@ public function __construct(
         $description,
         array $tags,
         BackgroundNode $background = null,
+        array $rules,
```
Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Node/FeatureNode.php:80]: [BC] we should add this as the last parameter and make it optional (with a default)

Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Node/FeatureNode.php:80]: I have a slight concern that we lose the ordering of scenarios+rules here by having them as separate arguments. (The main concern is the impact on Behat's formatters).

Perhaps we should be passing an array of Scenarios+Rules? Needs some thought

File: src/Behat/Gherkin/Node/FeatureNode.php [hunk 3]
```diff
@@ -185,6 +192,14 @@ public function getBackground()
         return $this->background;
     }
 
+    /**
+     * @return RuleNode[]
+     */
+    public function getRules(): array
```
Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Node/FeatureNode.php:198]: [BC] We need to consider how a consumer will know this method is available. Maybe we need an interface to hint it? I slightly prefer

```php
if ($featureNode instanceof FeatureRulesNode) {
```

vs

```php
if(method_exists($featureNode, 'getRules')){
```

File: tests/Behat/Gherkin/Keywords/KeywordsTest.php [hunk 4]
```diff
@@ -100,6 +100,7 @@ public function translationTestDataProvider()
                     ,
                     array(),
                     $background,
+                    [], // TODO: rules
```
Comment (ciaranmcnulty) [inline tests/Behat/Gherkin/Keywords/KeywordsTest.php:103]: I guess this should be resolved before merge

**Review by ciaranmcnulty — COMMENTED** [review]
It's really good that you've used the cucumber tests to guide things - your comment about the Scenario Outline difference is captured in #153 

I've left a few comments, the ones marked [BC] would allow us to do this as a minor release, I'm not sure if that's the right approach or not yet

**Comment by ciaranmcnulty** [comment]
@pamil @stof we need to consider how Behat will adapt to this change. It'll impact tagging, running and a lot more.

  1. We could do it in a minor version of behat/gherkin with BC layers, then Behat can do some conditional logic to see if the parser supports rules
     * Quite a 'safe' option
     * Could enable a situation where a user has a Rule-aware Gherkin but not a Rule-aware Behat and therefore rules aren't a parse error, but are ignored by the runner
     
  2. We could do it in a major, expand Behat's supported versions, and have similar conditional logic
     * Guarantees Behat understands rules if they're parsed
     * Probably messier conditional logic in Behat if we didn't put in a BC layer
  
  3. We could do it in a major and bump Behat's dependency overnight
     * Guarantees Behat understands rules if they're parsed
     * No BC issues
     * Could conflict on codebases that use e.g. Codeception+Behat (does that happen?)

File: src/Behat/Gherkin/Node/FeatureNode.php (hunk 3 above)
Comment (stof) [reply src/Behat/Gherkin/Node/FeatureNode.php:198]: Well, if a consumer does not want to have to check for existence, the easiest way is to bump their min version of behat/gherkin.

To me, there is no need for an interface (which would not be allowed to be removed without a BC break, and there is no right way to trigger deprecations for instanceof checks relying on deprecated interfaces, only static analysis can do it)

**Comment by unkind** [comment]
I found inconsistency with https://github.com/cucumber/common:

```gherkin
Feature: Some rules

  Background:
    Given fb

  Rule: A
    The rule A description

    Background:
      Given ab

    Example: Example A
      Given a
```

That parser doesn't remove indentation of description:

```json
"description":"    The rule A description"
```

**Comment by unkind** [comment]
#245 is just better way for this.

**Pull request closed without merging**